### PR TITLE
Multi-server client.

### PR DESCRIPTION
**Pull request by jsuresh**
Implemented by creating a new Client implementation, which makes writing
tests easier as well.

### CODE REVIEW

File: application.go [hunk 1]
```diff
@@ -49,7 +51,11 @@ type ApplicationDef struct {
 	HealthEndpoint string
 
 	// e.g. user@host  (no path)
-	Targets map[string]*Target
+	Targets map[TargetName]*Target
+
+	// Maps a name to a list of targets. Camu will then perform all
```
Comment (latos) [inline application.go]: Camu**s**

File: application.go [hunk 2]
```diff
@@ -94,7 +100,19 @@ func ApplicationFromConfig(isClient bool, file string) (Application, error) {
 		}
 
 		if !foundTarget {
-			return errMsg("No targets specified")
+			return errMsg("No single targets specified (need at least one)")
```
Comment (latos) [inline application.go]: "No server targets defined"

or s/server/regular/  "single" is ambiguous.

or "No 'Targets' entries defined" - probably the clearest

File: application.go [hunk 3]
```diff
@@ -94,7 +100,19 @@ func ApplicationFromConfig(isClient bool, file string) (Application, error) {
...
+		}
+
+		for name, group := range def.GroupTargets {
+			if _, ok := def.Targets[name]; ok {
+				return errMsg("%s appears as a target name in both GroupTargs and single Targets (target names must be unique across both maps)")
```
Comment (latos) [inline application.go]: s/single/<your-choice-above>/

wrap long line

File: application.go [hunk 4]
```diff
@@ -94,7 +100,19 @@ func ApplicationFromConfig(isClient bool, file string) (Application, error) {
...
+			}
+
+			for _, name := range group {
+				if _, ok := def.Targets[name]; !ok {
+					return errMsg("Expected %s to appear as an entry in 'Targets'. All target names in a target group list must be specified as a stand-alone target", name)
```
Comment (latos) [inline application.go]: wrap long line

File: client.go [hunk 5]
```diff
@@ -41,48 +41,56 @@ type SingleServerClient struct {
 	isLocalTest bool
 }
 
-func NewClientImpl(deployFile string, targetName string, isLocalTest bool) (*SingleServerClient, error) {
+// Client which communicates with multiple underlying servers at once. Used if
+// the target is actually a group of servers
+type MultiServerClient struct {
```
Comment (latos) [inline client.go]: MultiTargetClient?  "...ServerClient" sounds a bit confused :P 

Or just call this Client, and rename Client to TargetClient or SingleTargetClient.  Otherwise you should rename the "NewClient" to something else.

Comment (latos) [inline client.go]: Sorry, I forgot "Client" is the name of the interface.  Perhaps still use the term "Target" instead of "Server" everywhere.

File: client.go [hunk 6]
```diff
@@ -271,3 +279,98 @@ func (c *SingleServerClient) Shutdown() {
...
+			return nil, err
+		} else {
+			deploys = append(deploys, deploysForServer...)
+		}
+	}
+
+	return deploys, nil
+}
+
+func (c *MultiServerClient) KillUnknownProcesses() {
```
Comment (latos) [inline client.go]: I'm not sure what the purpose of this method is (same with shutdown).  I think too dangerous - if you really want to kill everything, log into the server.

It *would* be very convenient, however, to have a "killpid" command that kills by pid - easy way to clean up random cruft processes that we can see in the port range without having to log in, and is clear in how it works.

Comment (latos) [inline client.go]: (ack that this is unrelated to your change & this method's been there for a year)

File: client.go [hunk 7]
```diff
@@ -41,48 +41,56 @@ type SingleServerClient struct {
...
 	}
 
-	target := app.Target(targetName)
-	if target == nil {
-		return nil, fmt.Errorf("Invalid target: '%s'", targetName)
-	}
+	// Create all SingleServerClients
+	clients := []Client{}
+	for _, target := range app.Targets(targetName) {
+		localPort := target.Base
```
Comment (latos) [inline client.go]: I think there should be a different local port for each client (won't this not work with more than one client?)

In fact, why does it even need to be the same as the Base (this is probably why we have port conflicts in the first place - presumably a leftover from a very early version of camus)  - assuming I still remember correctly what these fields are for.

**Comment by latos** [comment]
howdy

File: application.go (hunk 1 above)
Comment (jsuresh) [inline application.go]: done

File: application.go (hunk 2 above)
Comment (jsuresh) [inline application.go]: done

File: application.go (hunk 3 above)
Comment (jsuresh) [inline application.go]: done

File: application.go (hunk 4 above)
Comment (jsuresh) [inline application.go]: done

File: client.go (hunk 6 above)
Comment (jsuresh) [inline client.go]: Shutdown is used by the integration test code (so we need it as part of the API), however, it should be removed from the terminal client: https://github.com/helix-collective/camus/issues/49

Agree that killUnknownProcesses is dangerous and should be removed altogether. I've raised a ticket for this: https://github.com/helix-collective/camus/issues/48

File: client.go (hunk 5 above)
Comment (jsuresh) [inline client.go]: Done s/Server/Target/g

File: client.go (hunk 7 above)
Comment (jsuresh) [inline client.go]: Each 'Target' is a new machine, so we don't get port clashes.

It doesn't need to be the same as the base, however, clients do need to know what port camus is running on. I'm guessing using the 'base' property for 2 things (the port camus listens on, and the port camus starts provisioning from) makes life easier in the early days?

Regardless - Not entirely clear as to what changes you are proposing in this particular case. If you would like to continue the discussion, may I suggest moving this over to a github issue.

**Comment by jsuresh** [comment]
Closed and merged into master, fixups implemented as a follow up comment: https://github.com/helix-collective/camus/commit/9afe57ed0c3595064e1e5c2b598baf4ee61188e0
